### PR TITLE
intel-llvm: init at unstable-2025-11-14

### LINKED ISSUE #245031: Package request: dpc++/c++
**Project description**

Intel oneAPI DPC++/C++ Compiler

Refer: [ArchLinux](https://archlinux.org/packages/extra/x86_64/intel-oneapi-dpcpp-cpp/)

Related packages:

- [mkl](https://search.nixos.org/packages?query=mkl)
- [intel-ocl](https://search.nixos.org/packages?query=mkl)
- [intel-compute-runtime](https://search.nixos.org/packages?query=intel-compute-runtime)

**Metadata**

* homepage URL: https://software.intel.com/content/www/us/en/develop/tools/oneapi.html
* source URL: <https://apt.repos.intel.com/oneapi/pool/main/intel-oneapi-dpcpp-cpp-2023.1.0-2023.1.0-46305_amd64.deb>
* license: unfree
* platforms: unix, linux, darwin

### PR DESCRIPTION
**Pull request by kilyanni**
Adds the Intel LLVM fork for SYCL compilation. This PR includes the compiler as well as an `stdenv` for making nix-derivations that use it.

Some notes:

- Upstreaming the `gnu-install-dirs.patch` patch wont make much sense until the [equivalent vanilla-LLVM patches](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/compilers/llvm/22/llvm/gnu-install-dirs.patch) are upstreamed (and reach the Intel Fork). The `sycl-jit-exclude-cmake-files.patch` only makes sense with the GNU one already applied, so that neither.
- This is packaged at an unstable commit instead of the last release, because several packaging-related changes are not yet in the release version. Most notably the dependency lookup changes from intel/llvm/issues/19635
- I've successfully used the stdenv to build large SYCL codebases, such as `oneDNN`. As far as I can tell, everything seems to work as intended.

Closes #245031
Related to #367722

## Things done

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by blenderfreaky** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 470035`
Commit: `87221ed18bcd8197bd02d8eb1e9ff2aac7fb429c`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>intel-llvm</li>
  </ul>
</details>

**Comment by kuflierl** [comment]
Finally we actually have intel-llvm. I can not even begin with how many hours I have tried to get the build system to handle offline build resources by reading through those CMake files. Thanks for getting this done (and partially upstreamed).

I will review this when I have a bit of time.

**Review by izelnakri — APPROVED** [review]

File: pkgs/by-name/in/intel-llvm/unified-runtime.nix [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    paths = with rocmPackages; [
+      clr
+      rocm-comgr
+      hsakmt
+    ];
+  };
+in
+stdenv.mkDerivation (finalAttrs: {
+  name = "unified-runtime";
+  inherit version;
```
Comment (kuflierl) [inline pkgs/by-name/in/intel-llvm/unified-runtime.nix]: version isn't used anyware else, why is it in a let expr?
```suggestion
  rocmtoolkit_joined = symlinkJoin {
    name = "rocm-merged";

    paths = with rocmPackages; [
      clr
      rocm-comgr
      hsakmt
    ];
  };
in
stdenv.mkDerivation (finalAttrs: {
  name = "unified-runtime";
  version = "0.12.0";
```

**Review by kuflierl — CHANGES_REQUESTED** [review]
Although i couldn't look into it deeply (especially gnu-install-dirs.patch), it seems everything looks fine (and works fine according to a few tests) except for a couple of nits.

What errors do you really experience with the built in tests?

Also you will need to squash those commits.

File: pkgs/by-name/in/intel-llvm/unified-runtime.nix (hunk 1 above)
Comment (blenderfreaky) [reply pkgs/by-name/in/intel-llvm/unified-runtime.nix]: Good catch, that's probably a left-over from an earlier attempt ^^'

**Comment by blenderfreaky** [comment]
I'm gonna have to start a fresh build to tell you what tests exactly fail, but it was mostly (if not entirely) tests in Clang.
All the LLVM tests passed. It's just because of the weird trick with the `ccWrapperStub` that the tests get a clang that behaves slightly differently than they'd expect (and there're many tests that care about those edge cases).

I tried a bunch of different ways to get `lit` to use the unwrapped clang, but couldn't ever get it to work properly.
The simplest way around that would probably be to not build `libdevice` alongside the rest, but I'm not sure it supports standalone builds (as most other Intel-authored LLVM packages don't either)

Also I'm gonna squash everything once the review is done, just keeping it for now to avoid the force-push mess

File: pkgs/by-name/in/intel-llvm/unwrapped.nix [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+  #       and other acceleration is already selected, disable it by default.
+  levelZeroSupport ? !(cudaSupport || rocmSupport),
+  nativeCpuSupport ? true,
+  enableManpages ? true,
+}:
+let
+  version = "unstable-2025-11-14";
+  date = "20251114";
+  llvmMajorVersion = "22";
+  src = fetchFromGitHub {
```
[the remark this answers is not in the document]
Comment (blenderfreaky) [reply pkgs/by-name/in/intel-llvm/unwrapped.nix]: This is actually somewhat non-trivial, as `unified-runtime` a bit below in the let block also depends on src.
It could all be moved inside, but even then overriding would still be finnicky because of the wrapping. I suppose something could probably be done via `makeScope` or `makeOverrideable`?

The vanilla LLVM derivation has `src` and co passed in as an input inside the derivation, something similar could work here maybe?

**Comment by blenderfreaky** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 470035 --additional-package pkgsRocm.intel-llvm`
Commit: `f4f6c98b51f9e2ca30902889badfa05ff3963d7e`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>intel-llvm</li>
    <li>pkgsRocm.intel-llvm</li>
  </ul>
</details>

**Comment by kuflierl** [comment]
You accidentally did a typo in your last commit "refator"  ^_^

**Review by kuflierl — APPROVED** [review]
Save for that nit I would say LGTM

**Comment by blenderfreaky** [comment]
Squashed everything into one commit, thanks for all the suggestions!

**Comment by blenderfreaky** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 470035 --additional-package pkgsCuda.intel-llvm --additional-package pkgsRocm.intel-llvm`
Commit: `f7ca69b03186270ccadbcfab317d9e10a342fcb8`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 3 packages built:</summary>
  <ul>
    <li>intel-llvm</li>
    <li>pkgsCuda.intel-llvm</li>
    <li>pkgsRocm.intel-llvm</li>
  </ul>
</details>

File: pkgs/by-name/in/intel-llvm/unwrapped.nix [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+  pkgsStatic,
+  hwloc,
+  emhash,
+  level-zero,
+  opencl-headers,
+  libxml2,
+  libedit,
+  llvmPackages,
+  # Ideally this would be the LLVM with the same major version
+  # as the LLVM built here (22). However LLVM 22 is packaged
```
Comment (RossComputerGuy) [inline pkgs/by-name/in/intel-llvm/unwrapped.nix]: LLVM 22 exists as it's own set as of this week.

File: pkgs/by-name/in/intel-llvm/unwrapped.nix [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+
+  buildInputs = [
+    spirv-tools
+    libxml2
+    hwloc
+    emhash
+    parallel-hashmap
+    pkgsStatic.zstd
+  ]
+  ++ unified-runtime.buildInputs;
```
Comment (RossComputerGuy) [inline pkgs/by-name/in/intel-llvm/unwrapped.nix]: If there's dependencies needed from that, they should be in unified-runtime's `propagatedBuildInputs`. Then you'll only need to add unified-runtime to the build inputs here.

**Review by RossComputerGuy — DISMISSED** [review]
This is a decent start. I took a cursory glance and have some changes. I do heavily recommend using our existing LLVM infrastructure in nixpkgs to assist.

**Comment by kuflierl** [comment]
> Do we really need more, vendor specific LLVMs?

Yes, it would be nice if we could get Intel specific GPU code running. Especially for blender. 

> Why can't they upstream this?

Don't quote me on this but GPU stuff generally isn't upstreamed due to gpus being super vendor specific. Ask upstream.

File: pkgs/by-name/in/intel-llvm/unwrapped.nix [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+    graphviz-nox
+  ];
+
+  buildInputs = [
+    spirv-tools
+    libxml2
+    hwloc
+    emhash
+    parallel-hashmap
+    pkgsStatic.zstd
```
[the remark this answers is not in the document]
Comment (blenderfreaky) [reply pkgs/by-name/in/intel-llvm/unwrapped.nix]: Their CMake explicitly pulls in the static version. Not sure why they do that, but it means (barring patching) it won't make it past configure without.
Also should be noted that the build needs to output static libraries currently due to upstream issues: https://github.com/intel/llvm/issues/19060

File: pkgs/by-name/in/intel-llvm/unwrapped.nix [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+      cudaSupport
+      rocmSupport
+      rocmGpuTargets
+      nativeCpuSupport
+      ;
+  };
+in
+# Tip: This build plays nice with ccacheStdenv.
+#      Replace stdenv here to make debugging less tedious.
+stdenv.mkDerivation (finalAttrs: {
```
[the remark this answers is not in the document]
Comment (blenderfreaky) [reply pkgs/by-name/in/intel-llvm/unwrapped.nix]: I think it might be possible, but their current CMake & runtime rather strongly assumes monolithic builds

What exactly are the benefits of splitting here, btw?

In the case of this LLVM fork, you rarely (if ever) want to pull in specific parts, you basically only care about their `clang` for the `-fsycl` support. Basically everything else in it is just a support library for their `clang`, and I'm not aware of any packages that depend on their libllvm or similar

[the remark this answers is not in the document]
Comment (RossComputerGuy) [reply pkgs/by-name/in/intel-llvm/unwrapped.nix]: Splitting is good since it helps with cross compilation. It also makes building multiple derivations a bit nicer. There's also the reduction of people's closures.

You can also leverage the existing LLVM infra which makes the size of this PR smaller and more maintainable.

**Comment by kilyanni** [comment]
The core changes needed are mostly the PRs by sarnex from https://github.com/intel/llvm/issues/19635.
These do not seem to have been merged into 6.3.0 (their main `sycl` branch and releases seem to be rather divergent).

These changes affect the way dependencies are looked up and thus change the derivation quite a bit.

I suppose I could cherry pick the specific PRs ontop of a release?

Tangentially; as UMF recently got a new release (#493489) and as did L0, I could update this to a newer unstable version now. Though it seems the bleeding edge currently breaks some builds (e.g. due to them removing `syclcompat`)

But either way, merging is blocked on RossComputerGuy at the moment (see the comment chain from the review above, I wrote a small breakdown yesterday regarding some issues with the requested changes)

File: pkgs/by-name/in/intel-llvm/unwrapped.nix [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+    inherit unified-runtime;
+
+    # This is for easily referencing version-compatible LLVM libraries
+    # and tools that aren't built in this derivation,
+    # as well as nix tooling, such as the stdenv.
+    baseLlvm = llvmPackages_22;
+
+    inherit llvmMajorVersion;
+
+    updateScript = nix-update-script { };
```
Comment (doronbehar) [inline pkgs/by-name/in/intel-llvm/unwrapped.nix]: FYI, this does nothing, see:

- https://github.com/NixOS/nixpkgs/pull/502056

File: pkgs/by-name/in/intel-llvm/unwrapped.nix [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+
+    # This hardening option causes compilation errors when compiling for amdgcn, spirv and others
+    hardeningUnsupportedFlags = [ "zerocallusedregs" ];
+  };
```
Comment (doronbehar) [inline pkgs/by-name/in/intel-llvm/unwrapped.nix:305]: Most people are used to having `meta` be the last argument of `stdenv.mkDerivation`, and `passthru` above it.

File: pkgs/by-name/in/intel-llvm/unwrapped.nix (hunk 4 above)
Comment (doronbehar) [reply pkgs/by-name/in/intel-llvm/unwrapped.nix:129]: > `unified-runtime` is a bit of a pseudo-package; it's somewhat independent but not really. Their LLVM CMake doesn't allow linking to an externally built UR, instead it's built in-tree (which is why `unified-runtime.cmakeFlags` also gets appended for example).
> 
> The reason I gave it a separate derivation is mostly for separation of concerns, as most of the GPU-driver related code is in there, which is whats most likely to break during any updates (both to Intel LLVM and driver packaging).
> 
> I'm not sure if there's any external users of UR at the moment

I guess this is a fine argument for the meantime, but please consider maybe dismantling `unified-runtime` in the future.

File: pkgs/by-name/in/intel-llvm/unwrapped.nix (hunk 5 above)
[the remark this answers is not in the document]
Comment (doronbehar) [reply pkgs/by-name/in/intel-llvm/unwrapped.nix]: @kilyanni using `pkgsStatic` here will cause a mixture of musl & GNU stdlib derivations, which is considered dangerous (so I heard in the past). Have you considered using an overriden derivation of `zstd` with `--disable-shared` and `--enable-static` or something alike?

File: pkgs/by-name/in/intel-llvm/unwrapped.nix [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+stdenv.mkDerivation (finalAttrs: {
+  pname = "intel-llvm";
+
+  inherit src version commitDate;
+
+  outputs = [
+    "out"
+    "lib"
+    "dev"
+    "python"
```
Comment (doronbehar) [inline pkgs/by-name/in/intel-llvm/unwrapped.nix:94]:
```suggestion
    "python"
    # Not adding (conditionally) a "man" output here to avoid complexity in the
    # wrapper, that is not worth it for a few Kb.
```

**Review by doronbehar — COMMENTED** [review]
Mostly looks good I guess, but I skimmed over most of the comments. @RossComputerGuy's comments were good so it is a shame they stopped replying and abandoned the review of this PR.

File: pkgs/by-name/in/intel-llvm/unwrapped.nix (hunk 7 above)
Comment (kilyanni) [reply pkgs/by-name/in/intel-llvm/unwrapped.nix]: Oh wow, really? I never knew ^^'

Thanks for the tip!

File: pkgs/by-name/in/intel-llvm/unwrapped.nix (hunk 5 above)
[the remark this answers is not in the document]
Comment (kilyanni) [reply pkgs/by-name/in/intel-llvm/unwrapped.nix]: Ah, I wasn't aware that that pulls in musl
I think `zstd.override { static = true; }` should achieve the same effect as far as this package is concerned

Running the full build with it right now, it passed configure without issues at the least :+1:

**Comment by kilyanni** [comment]
^ the `unified-memory-framework` commit fixes something that broke between updates and only affected the version used by this derivation (with `.override { useJemalloc = false; }`)

**Comment by kilyanni** [comment]
Fixed the vet failure and some issues that popped up from rebasing, also squashed all the commits down

All builds pass locally (~~with ccacheStdenv, 'proper' build of all 3 variants would take like 6h or something~~)

@doronbehar I addressed the nits and replaced `pkgsStatic.zstd` with `zstd.override { statice = true; }`. Can you take a look and see if this addresses your issues?

**Comment by kilyanni** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 470035 --additional-package pkgsRocm.intel-llvm --additional-package pkgsCuda.intel-llvm`
Commit: `635d93e5d440e91fd4836948e5b1c817354d7ae1`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>intel-llvm</li>
    <li>pkgsCuda.intel-llvm</li>
    <li>pkgsRocm.intel-llvm</li>
    <li>unified-memory-framework</li>
  </ul>
</details>

File: pkgs/by-name/in/intel-llvm/unwrapped.nix [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+    parallel-hashmap
+    # Static ZSTD is required by sycl, see sycl/source/CMakeLists.txt:163
+    # This may be related to the fork not supporting building shared libraries;
+    #  https://github.com/intel/llvm/issues/19060
+    (zstd.override { static = true; })
+  ]
+  ++ unified-runtime.buildInputs;
+
+  propagatedBuildInputs = [
+    zstd
```
Comment (doronbehar) [inline pkgs/by-name/in/intel-llvm/unwrapped.nix]: Shouldn't this be `(zstd.override { static = true; })`? If so, maybe create a `zstdStatic` above under `let`-`in`. I wonder in general: Why do you need to propagate it?

File: pkgs/by-name/in/intel-llvm/unwrapped.nix [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+  cmakeFlags = [
+    (lib.cmakeBool "LLVM_INSTALL_UTILS" true)
+
+    (lib.cmakeBool "LLVM_BUILD_TESTS" false)
+    (lib.cmakeBool "LLVM_INCLUDE_TESTS" false)
+    (lib.cmakeBool "MLIR_INCLUDE_TESTS" false)
+    (lib.cmakeBool "SYCL_INCLUDE_TESTS" false)
+
+    (lib.cmakeFeature "LLVM_ENABLE_ZSTD" "FORCE_ON")
+    (lib.cmakeBool "LLVM_USE_STATIC_ZSTD" true)
```
Comment (doronbehar) [inline pkgs/by-name/in/intel-llvm/unwrapped.nix]: I assume that using regular `zstd` here and setting it to `false` is something you tested, and that doesn't work. Could you maybe expand on that in a comment?

**Review by doronbehar — COMMENTED** [review]
Looking better, thanks for the corporation! 2 more small comments.

File: pkgs/by-name/in/intel-llvm/unwrapped.nix (hunk 10 above)
Comment (kilyanni) [reply pkgs/by-name/in/intel-llvm/unwrapped.nix]: Nice catch ^^

I'm actually not fully sure why I propagated it. I think it had to do with me trying to use the non-static version in an earlier version, then forgetting to remove it.

The build works fine with (static) `zstd` only in `buildInputs`, so I removed the other two

File: pkgs/by-name/in/intel-llvm/unwrapped.nix (hunk 11 above)
Comment (kilyanni) [reply pkgs/by-name/in/intel-llvm/unwrapped.nix]: It seems this is actually already set automatically by some of Intels CMake, so it's redundant to set manually. I removed the explicit `LLVM_USE_STATIC_ZSTD`.

Now `zstd` only appears in `buildInputs`, so I slightly expanded the comment there to explain

**Review by doronbehar — APPROVED** [review]
Looks good to me. I trust the author to have tested this, and thanks to the other positive reviews I think this should be good enough to go in.

**Comment by kilyanni** [comment]
Thanks!

File: pkgs/by-name/in/intel-llvm/unwrapped.nix [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+    )
+
+    # We eval because flags is separated as shell-escaped strings.
+    # We can't just split by space because it may contain escaped spaces,
+    # so we just let bash handle it.
+    # NOTE: We prepend, so that flags we set manually override what the build script does.
+    eval "prependToVar cmakeFlags $flags"
+
+    # Remove the install prefix flag
+    cmakeFlags=(''${cmakeFlags[@]/-DCMAKE_INSTALL_PREFIX=$NIX_BUILD_TOP\/source\/build\/install})
```
Comment (ShamrockLee) [inline pkgs/by-name/in/intel-llvm/unwrapped.nix:202]: `cmake`'s `cmakeConfigurePhase` prepends `-DCMAKE_INSTALL_PREFIX`.

https://github.com/NixOS/nixpkgs/blob/bc071daf4c377dd821398672fafe269d5c654ce8/pkgs/by-name/cm/cmake/setup-hook.sh#L26-L28

This line doesn't work as you commented, as you don't specify `-DCMAKE_INSTALL_PREFIX` in `cmakeFlags`, and there's no other hook adding such flag to `cmakeFlags`.

To prevent `cmakeConfigurePhase` from specifying `CMAKE_INSTALL_PREFIX`, you can set `{ dontAddPrefix = true; }`.
This is probably unneeded, since the package already builds.

Comment (doronbehar) [reply pkgs/by-name/in/intel-llvm/unwrapped.nix:202]: @ShamrockLee in your comment, is there an action suggested to the PR author? Do you think something in this package could have been more accurate?

Comment (ShamrockLee) [reply pkgs/by-name/in/intel-llvm/unwrapped.nix:202]: These lines are effectively dead code and can be removed.

Comment (ShamrockLee) [reply pkgs/by-name/in/intel-llvm/unwrapped.nix:202]: I take back what I said.

Some CMake Flags are injected from the output of  `buildbot/configure.py`, which could produce (wrong) `-DCMAKE_INSTALL_PREFIX` flags.

Removing these lines causes build failure.

Comment (kilyanni) [reply pkgs/by-name/in/intel-llvm/unwrapped.nix:202]: Ah, yeah, that was it. The script sets its own default and overriding it was easier than filtering it out.

Making a note to myself to make the comment a bit more explanatory next time I PR some changes to this package
